presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectStreamReader.java

```diff
@@ -144,47 +149,69 @@ public Block readBlock(Type type)
-            if (dataStream == null) {
-                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
-            }
-            data = dataStream.next(toIntExact(totalLength));
+        if (dataStream == null) {
+            throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
         }
 
-        Slice[] sliceVector = new Slice[nextBatchSize];
-
-        int offset = 0;
-        for (int i = 0; i < nextBatchSize; i++) {
-            if (!isNullVector[i]) {
-                int length = lengthVector[i];
-                Slice value = Slices.wrappedBuffer(data, offset, length);
-                if (isVarcharType(type)) {
-                    value = truncateToLength(value, type);
-                }
-                if (isCharType(type)) {
-                    value = trimSpacesAndTruncateToLength(value, type);
-                }
-                sliceVector[i] = value;
-                offset += length;
+        // allocate enough space to read
+        byte[] data = new byte[toIntExact(totalLength)];
+        Slice slice = Slices.wrappedBuffer(data);
+
+        // truncate string and update offsets
+        for (int i = 0; i < currentBatchSize; i++) {
+            // the previously computed offsets are original offsets without truncation
+            // update the offsets in this loop to be the truncated ones
+            offsetVector[i + 1] = offsetVector[i];
+            if (isNullVector[i]) {
+                continue;
+            }
+            int offset = offsetVector[i];
+            int length = lengthVector[i];
+
+            // read data without truncation
+            dataStream.next(offset + length, data, offset);
+
+            // calculate truncated length
+            int truncatedLength;
+            if (isVarcharType(type)) {
+                VarcharType varcharType = (VarcharType) type;
+                // code point is at least 1 byte; therefore it cannot exceed the slice length
+                truncatedLength = getTruncatedLength(slice, offset, length, varcharType.isUnbounded() ? length : varcharType.getLengthSafe());
+            }
+            else if (isCharType(type)) {
+                int trimmedLength = sliceLengthWithoutTrailingSpaces(slice, offset, length);
+                verify(trimmedLength >= 0);
+                truncatedLength = getTruncatedLength(slice, offset, trimmedLength, ((CharType) type).getLength());
+            }
+            else {
+                truncatedLength = length;
             }
-        }
 
-        readOffset = 0;
-        nextBatchSize = 0;
+            // adjust offsets with truncated length
+            offsetVector[i + 1] += truncatedLength;
+        }
 
-        return new SliceArrayBlock(sliceVector.length, sliceVector);
+        return new VariableWidthBlock(currentBatchSize, slice, offsetVector, isNullVector);
```

Comment:
comment that this can lead to over-retention. But it's not a big deal given the unlikeliness that this happens.